[all java] Add Java module support

SamCarlberg opened the pull request:
Resolves #7236 

Non-modular Java dependencies (ejml, opencv) need a gradle plugin to add a module-info file to the JAR. I'm not sure if this is the correct approach, since those JARs won't be repacked and published for user programs to consume. Will need to test the artifacts to see if user programs will be broken

KangarooKoala commented on wpilibNewCommands/src/main/java/module-info.java:
```diff
@@ -0,0 +1,14 @@
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+module wpilib.commands2 {
```
Open discussion: `wpilib.commands` or `wpilib.commands2`? The module names already differ from the package names.

calcmogul replied:
`commands2` makes sense if we plan on making a `commands3`, unless there's some way to version it separately. If we can use vendordep versioning and prevent users from installing two versions at once, we could just use `commands`.

SamCarlberg replied:
`commands2` makes the most sense. We'll have a commands3 in 2027, and each version is essentially a completely different library so reusing the same module name wouldn't be practical (or likely even possible)

calcmogul replied:
That argument doesn't make sense to me. WPILib 2008 and 2024 look completely different, yet we still call it wpilibj. Libraries going through a major API refactor are why major version numbers exist, so we should be using the major version for that instead of the library name itself.

SamCarlberg replied:
Because wpilib was changed gradually over time, without multiple incompatible versions concurrently being available for use. Commands v3 will probably be available side by side with v2 during a transition period

calcmogul replied:
That's the case for most libraries though, and they don't name their libraries like that. Giving them different names would allow users to use them both simultaneously, which we should explicitly disallow. Allowing that for v1 and v2 was a mistake.

SamCarlberg commented on epilogue-processor/src/main/java/edu/wpi/first/epilogue/processor/EpilogueGenerator.java:
Yes. The generated file is in the robot project, which would result in the `edu.wpi.first.epilogue` package being split between the `wpilib.epilogue` module and the robot program's module. Generating the Epilogue class in its own package fixes that.

I was also considering moving the generated Epilogue file to be in the same package as the main robot class, but that doesn't work if there are multiple robot classes in the same project

SamCarlberg committed:
Add module-info files to all libraries

SamCarlberg committed:
Use modular OpenCV build and remove avaje-jsonb-inject-plugin

avaje-jsonb-inject-plugin is incorrectly listed as a dependency of io.avaje:jsonb, and causes runtime crashes if the `--add-modules ALL-MODULE-PATH` option is used because it forces the JVM to try to load io.avaje.inject, which is a dependency injection library completely unrelated to JSON handling.

SamCarlberg committed:
Move `Preferences` class to org.wpilib.preferences package

Avoids splitting the `org.wpilib.util` package

SamCarlberg committed:
Generate Epilogue class in `org.wpilib.epilogue.generated` package

To avoid splitting the `org.wpilib.epilogue` package between epilogue-runtime and generated user code

SamCarlberg committed:
Add allwpilib-java umbrella project

This provides a single module definition so projects can use `import module wpilib` instead of separate imports for `wpilib.core`, `wpilib.util`, `wpilib.hal`, et cetera

SamCarlberg committed:
Bazel build updates

SamCarlberg committed:
Add generated sources to source sets

Allows docs to build

SamCarlberg committed:
Update runtime JVM flags to use module names instead of ALL-UNNAMED

SamCarlberg committed:
Make developerRobot modular

SamCarlberg committed:
Add umbrella project as a dependency for wpilibjExamples

SamCarlberg committed:
Improve module-info documentation

SamCarlberg committed:
Support modules in doc generation

Instead of ignoring modularity, docs are modules-first and module-info javadocs will actually appear in the API documentation

SamCarlberg committed:
Merge branch 'main' into java-modules

pjreiniger reviewed (commented):
The bazel portions look fine. 

Its unfortunate that I couldn't find a workaround for the header jars, but we can circle back to that or make in issue in rules_java

auscompgeek commented:
> Its unfortunate that I couldn't find a workaround for the header jars, but we can circle back to that or make in issue in rules_java

Could we transition that flag just for the targets it's a problem for, instead of globally disabling it?

pjreiniger commented:
> 
> Could we transition that flag just for the targets it's a problem for, instead of globally disabling it?

I think it be better to follow up with that, no need to block this PR. I have to imagine the disabling the header jars has less of an impact on build time compared to compiling one object file in wpimath.

SamCarlberg commented:
I imagine the header jars are more useful for very large Java projects at Google's scale. Our Java projects are relatively tiny and compile quickly, particularly in comparison with native libraries like wpimath.

auscompgeek commented:
I agree that ijars probably don't have much of an impact on uncached builds for us, however they would improve incremental builds with no API changes in between.

Happy for it to be a follow up though.

SamCarlberg committed:
Merge branch 'main' into java-modules

SamCarlberg committed:
Update wpimath module exports

The global .proto package was split into separate *.proto.detail packages

Add `shape` package and its subpackages

SamCarlberg committed:
Remove avaje-related specifications from wpilib.apriltag

SamCarlberg committed:
Update wpilib.gamefields and the example dependency on it

SamCarlberg committed:
Patch gamefields module in bazel build

SamCarlberg committed:
Fix JUnit in wpilibjExamples bazel build
